jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/internal/HttpStreamOverHTTP2.java

```diff
@@ -241,12 +241,19 @@ public Runnable onTrailer(HeadersFrame frame)
     private Content.Chunk createChunk(Stream.Data data)
     {
         DataFrame frame = data.frame();
-        if (frame.isEndStream() && frame.remaining() == 0)
-            return Content.Chunk.EOF;
-
+        ByteBuffer byteBuffer = frame.getData();
+        if (!byteBuffer.hasRemaining())
+        {
+            // We must NOT retain because we are not passing the
+            // ByteBuffer to the Chunk.
+            if (frame.isEndStream())
+                return Content.Chunk.EOF;
+            else
+                return Content.Chunk.EMPTY;
+        }
```

Comment:
again, can't we just let `from` do it's stuff rather than put this logic everywhere?